src/pip/_internal/metadata/__init__.py

```diff
@@ -60,6 +61,11 @@ def select_backend() -> Backend:
         from . import importlib
 
         return cast(Backend, importlib)
+
+    if sys.version_info >= (3, 12):
+        message = "Cannot use pkg_resources as metadata backend on Python 3.12+"
+        raise PipError(message)
+
```

Comment:
If a bistro does the wrong thing and this gets exposed to users; it's a fairly unfriendly error message. Thoughts on raising a `DiagnosticPipError` subclass here?

Reply:
It took me *way* too long to realise you meant "distro" rather than "bistro" :-) (I see you've corrected it now).

I'm not particularly sure from a brief check of the code what switching to `DiagnosticPipError` would do in practice, but +1 on making it clearer that this is a "something is broken, talk to your distro" message, rather than a "you did something wrong" message.

Reply:
It’s gonna be presented prettily with rich, optionally with colors depending on flags and terminal capabilities.

See https://github.com/pypa/pip/blob/main/tests/unit/test_exceptions.py#L284 for example outputs.

Reply:
Just to be clear, by this you mean a distro would remove/split `importlib.metadata`, right?

Reply:
No, this is if they configure pip to use pkg_resources on 3.12 as the metadata backend.

That's exactly the sort of confusion that a well-formed error message would avoid. :)

Reply:
I thought about this a bit more; maybe instead of failing, we should simply ignore this setting on 3.12+? This can only be set by one of the following two ways: an attribute in `importlib.metadata`, or the environment variable.

Since the attribute is set by the Python distributor, we can simply make the patch non-functional, and it should be caught by the distro developers. The environment variable is the more problematic, but since we don’t advertise it anywhere, it should only be set by either distro developers (who know what they’re doing anyway) or a user who had a problem (maybe a mis-handled **3.11** installation) and was informed of this misguided solution. Either way, the user likely does not expect the config to be effective on 3.12 anyway, and it’s probably beneficial if we do not make it so.